apps/server/src/modules/board/service/board-do-copy.service.ts

```diff
@@ -0,0 +1,239 @@
+};
+
+class RecursiveCopyVisitor implements BoardCompositeVisitorAsync {
+	resultMap = new Map<EntityId, CopyStatus>();
+
+	copyMap = new Map<EntityId, AnyBoardDo>();
+
+	constructor(
+		private readonly filesStorageClientAdapterService: FilesStorageClientAdapterService,
+		private readonly params: RecursiveCopyVisitorParams
+	) {}
+
+	async copy(original: AnyBoardDo): Promise<CopyStatus> {
+		await original.acceptAsync(this);
+
+		const result = this.resultMap.get(original.id);
+		/* istanbul ignore next */
+		if (result === undefined) {
+			throw new Error('nothing copied');
+		}
+		return result;
+	}
+
+	async visitColumnBoardAsync(original: ColumnBoard): Promise<void> {
+		await this.visitChildrenOf(original);
+
+		const copy = new ColumnBoard({
+			id: new ObjectId().toHexString(),
+			title: original.title,
+			context: original.context,
+			createdAt: new Date(),
+			updatedAt: new Date(),
+			children: this.getCopiesForChildrenOf(original),
+		});
+
+		this.resultMap.set(original.id, {
+			copyEntity: copy,
+			type: CopyElementType.COLUMNBOARD,
+			status: CopyStatusEnum.SUCCESS,
+			elements: this.getCopyStatusesForChildrenOf(original),
+		});
+		this.copyMap.set(original.id, copy);
+	}
+
+	async visitColumnAsync(original: Column): Promise<void> {
+		await this.visitChildrenOf(original);
+		const copy = new Column({
+			id: new ObjectId().toHexString(),
+			title: original.title,
+			children: this.getCopiesForChildrenOf(original),
+			createdAt: new Date(),
+			updatedAt: new Date(),
+		});
+		this.resultMap.set(original.id, {
+			copyEntity: copy,
+			type: CopyElementType.COLUMN,
+			status: CopyStatusEnum.SUCCESS,
+			elements: this.getCopyStatusesForChildrenOf(original),
+		});
+		this.copyMap.set(original.id, copy);
+	}
+
+	async visitCardAsync(original: Card): Promise<void> {
+		await this.visitChildrenOf(original);
+		const copy = new Card({
+			id: new ObjectId().toHexString(),
+			title: original.title,
+			height: original.height,
+			children: this.getCopiesForChildrenOf(original),
+			createdAt: new Date(),
+			updatedAt: new Date(),
+		});
+		this.resultMap.set(original.id, {
+			copyEntity: copy,
+			type: CopyElementType.CARD,
+			status: CopyStatusEnum.SUCCESS,
+			elements: this.getCopyStatusesForChildrenOf(original),
+		});
+		this.copyMap.set(original.id, copy);
+	}
+
+	async visitFileElementAsync(original: FileElement): Promise<void> {
+		const copy = new FileElement({
+			id: new ObjectId().toHexString(),
+			caption: original.caption,
+			children: [],
+			createdAt: new Date(),
+			updatedAt: new Date(),
+		});
+		const fileCopy = await this.filesStorageClientAdapterService.copyFilesOfParent({
+			source: {
+				parentId: original.id,
+				parentType: FileRecordParentType.BoardNode,
+				schoolId: this.params.originSchoolId,
+			},
+			target: {
+				parentId: copy.id,
+				parentType: FileRecordParentType.BoardNode,
+				schoolId: this.params.targetSchoolId,
+			},
+			userId: this.params.userId,
```

Comment:
Does that mean that all files in the target board will belong the same user?

Reply:
all file copies belong to the user that triggered the copy process, as is also the case for other copied files so far